lxd/storage/backend_lxd.go

```diff
@@ -0,0 +1,682 @@
+		if err == db.ErrNoSuchObject {
+			return fmt.Errorf("Source volume doesn't exist")
+		}
+
+		return err
+	}
+
+	// Use the source volume's config if not supplied.
+	if config == nil {
+		config = srcVolRow.Config
+	}
+
+	// Use the source volume's description if not supplied.
+	if desc == "" {
+		desc = srcVolRow.Description
+	}
+
+	// Check the supplied config and remove any fields not relevant for destination pool type.
+	err = b.driver.ValidateVolume(config, true)
+	if err != nil {
+		return err
+	}
+
+	// Create slice to record DB volumes created if revert needed later.
+	revertDBVolumes := []string{}
+	defer func() {
+		// Remove any DB volume rows created if we are reverting.
+		for _, volName := range revertDBVolumes {
+			b.state.Cluster.StoragePoolVolumeDelete("default", volName, db.StoragePoolVolumeTypeCustom, b.ID())
+		}
+	}()
+
+	// Create database entry for new storage volume.
+	err = VolumeDBCreate(b.state, b.name, volName, desc, db.StoragePoolVolumeTypeNameCustom, false, config)
+	if err != nil {
+		return err
+	}
+
+	revertDBVolumes = append(revertDBVolumes, volName)
+
+	if copySnapshots {
+		// If we are copying snapshots, retrieve a list of snapshots from source volume.
+		snapshots, err := VolumeSnapshotsGet(b.state, srcPoolName, volName, db.StoragePoolVolumeTypeCustom)
+		if err != nil {
+			return err
+		}
+
+		// Create a database entry and copy the volume for each snapshot.
+		for _, srcSnapshot := range snapshots {
+			// Convert the source snapshot volume name into the new snapshot volume name.
+			_, snapName, _ := shared.ContainerGetParentAndSnapshotName(srcSnapshot.Name)
+			newSnapshotName := fmt.Sprintf("%s%s%s", volName, shared.SnapshotDelimiter, snapName)
+
+			// Create database entry for new storage volume.
+			err = VolumeDBCreate(b.state, b.name, newSnapshotName, srcSnapshot.Description, db.StoragePoolVolumeTypeNameCustom, true, config)
+			if err != nil {
+				return err
+			}
+
+			revertDBVolumes = append(revertDBVolumes, newSnapshotName)
+		}
+	}
+
+	srcVol := srcPool.newVolume(drivers.VolumeTypeCustom, drivers.ContentTypeFS, srcVolName, nil)
+	targetVol := b.newVolume(drivers.VolumeTypeCustom, drivers.ContentTypeFS, volName, config)
+	err = b.driver.CreateVolumeFromCopy(targetVol, srcVol, copySnapshots, op)
+	if err != nil {
+		return err
+	}
+
+	revertDBVolumes = nil // Don't revert DB volumes.
+	return nil
+}
+
+// MigrateCustomVolume sends a volume for migration.
+func (b *lxdBackend) MigrateCustomVolume(conn *websocket.Conn, args migration.VolumeSourceArgs, op *operations.Operation) error {
+	vol := b.newVolume(drivers.VolumeTypeCustom, drivers.ContentTypeFS, args.Name, nil)
+	err := b.driver.MigrateCustomVolume(vol, conn, args, op)
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// CreateCustomVolumeFromMigration receives a volume being migrated.
+func (b *lxdBackend) CreateCustomVolumeFromMigration(conn *websocket.Conn, args migration.VolumeTargetArgs, op *operations.Operation) error {
```

Comment:
Ah yeah, we still need to figure out a wrapper ReadWriteCloser around websocket.Conn so we can have the migration layer convert from websocket.Conn to ReadWriteCloser